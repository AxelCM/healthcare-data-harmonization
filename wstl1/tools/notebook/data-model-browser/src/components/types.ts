// Copyright 2020 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Meta data describing a top level entity/resource within a data model.
 */
export interface TopEntityMeta {
  /**
   * The name of the entity (e.g. Patient)
   */
  name: string;

  /**
   * A detailed description of the entity.
   */
  description: string;
}

/**
 * Meta data describing all sub entities/fields within a data model.
 */
export interface SubEntityMeta {
  /**
   * Name of the sub entity/field.
   */
  name: string;

  /**
   * A flag indicating whether the field is required.
   */
  required: boolean;

  /**
   * The type of the sub entity. The type could either be a primitive in the
   * data model (e.g. string, number), or a reference to another complex type
   * (e.g. Date, Identifier in FHIR).
   */
  type: string;

  /**
   * A detailed description of the sub entity or field.
   */
  description: string;

  /**
   * The JSON schema definition of the sub entity.
   */
  // tslint:disable-next-line:no-any
  schema?: any;

  /**
   * A flag indicating whether the field contains a complex type which can be
   * further explored or a primitive type that can not be further explored.
   */
  clickable: boolean;
}

/**
 * A result of a search query.
 */
export interface SearchResult {
  /**
   * The path of the search result, including the associated data model.
   * For example, FHIR > Patient > address.
   */
  path: string;

  /**
   * The name of the top level resource associated with the search result.
   */
  resource: string;

  /**
   * The field associated with the search result. If the selected search result
   * was for a top level entity, then the field will be empty.
   */
  field: string;
}
