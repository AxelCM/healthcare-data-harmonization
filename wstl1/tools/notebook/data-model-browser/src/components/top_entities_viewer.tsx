/**
 * @jsx React.createElement
 * @jsxFrag React.Fragment
 */
// Copyright 2020 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// tslint:disable-next-line:enforce-name-casing
import * as React from 'react';

import {makeStyles} from '@material-ui/core/styles';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableRow from '@material-ui/core/TableRow';

import {InspectTooltip} from './inspect_tool_tip';
import {TopEntityTooltip} from './top_entity_tool_tip';
import {TopEntityMeta} from './types';

const localStyles = makeStyles({
  span: {
    cursor: 'pointer',
    '&:hover': {
      textDecoration: 'underline',
    },
  },
});

/**
 * Properties to customize the TopEntitiesViewer component.
 */
interface Props {
  /**
   * A list of all of the top level entites renderable by the component.
   */
  topEntities: TopEntityMeta[];

  /**
   * The name of the selected/active top level entity.
   */
  selected: string;

  /**
   * Callback invoked when a specific top level entity is inspected.
   */
  onInspect: (name: string) => void;

  /**
   * Callback invoked when a specific top level entity is selected.
   */
  onSelect: (name: string) => void;
}

/**
 * Component to visualize all the top level entites within a data model.
 */
// tslint:disable-next-line:enforce-name-casing React component.
export function TopEntitiesViewer(props: Props) {
  const [open, setOpen] = React.useState(false);
  const classes = localStyles();
  let keyIndex = 0;
  const handleTooltipClose = () => {
    setOpen(false);
  };
  return (
    <ClickAwayListener onClickAway={handleTooltipClose}>
      <Table>
        <TableBody>
          {props.topEntities.map(meta => {
            const isSelected = meta.name === props.selected;
            return (
              <InspectTooltip
                key={keyIndex++}
                title={TopEntityTooltip(meta)}
                open={isSelected && open}
                disableFocusListener
                disableHoverListener
                disableTouchListener
              >
                <TableRow
                  hover
                  key={meta.name}
                  onClick={e => {
                    setOpen(true);
                    props.onInspect(meta.name);
                  }}
                  selected={isSelected}
                >
                  <TableCell component="th" scope="row">
                    <span
                      className={classes.span}
                      onClick={e => {
                        props.onSelect(meta.name);
                      }}
                    >
                      {meta.name}
                    </span>
                  </TableCell>
                </TableRow>
              </InspectTooltip>
            );
          })}
        </TableBody>
      </Table>
    </ClickAwayListener>
  );
}
