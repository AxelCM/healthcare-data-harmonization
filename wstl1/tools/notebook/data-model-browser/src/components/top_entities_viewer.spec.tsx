// Copyright 2020 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// tslint:disable-next-line:enforce-name-casing
import * as React from 'react';
import {shallow} from 'enzyme';
import TableRow from '@material-ui/core/TableRow';
import {TopEntitiesViewer} from './top_entities_viewer';
import {TopEntityMeta} from './types';

describe('TopEntitiesViewer', () => {

  it('should render correctly', () => {
    const entities: TopEntityMeta[] = [
      {name: 'Test', description: 'Verbose details'},
    ];
    const selected = 'Test';
    const onInspect = jest.fn();
    const onSelect = jest.fn();

    const wrapper = shallow(
      <TopEntitiesViewer
        topEntities={entities}
        onInspect={onInspect}
        onSelect={onSelect}
        selected={selected}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the correct function on entity row click', () => {
    const entities: TopEntityMeta[] = [
      {name: 'Test', description: 'Verbose details'},
    ];
    const selected = 'Test';
    const onInspect = jest.fn();
    const onSelect = jest.fn();

    const wrapper = shallow(
      <TopEntitiesViewer
        topEntities={entities}
        onInspect={onInspect}
        onSelect={onSelect}
        selected={selected}
      />
    );
    wrapper.find(TableRow).simulate('click');
    expect(onInspect).toHaveBeenCalled();
    expect(onInspect).toBeCalledWith('Test');
    expect(onSelect).not.toHaveBeenCalled();
  });

  it('should call the correct function on entity cell click', () => {
    const entities: TopEntityMeta[] = [
      {name: 'Test', description: 'Verbose details'},
    ];
    const selected = 'Test';
    const onInspect = jest.fn();
    const onSelect = jest.fn();

    const wrapper = shallow(
      <TopEntitiesViewer
        topEntities={entities}
        onInspect={onInspect}
        onSelect={onSelect}
        selected={selected}
      />
    );
    wrapper.find('span').simulate('click');
    expect(onInspect).not.toHaveBeenCalled();
    expect(onSelect).toHaveBeenCalled();
    expect(onSelect).toBeCalledWith('Test');
  });
});
