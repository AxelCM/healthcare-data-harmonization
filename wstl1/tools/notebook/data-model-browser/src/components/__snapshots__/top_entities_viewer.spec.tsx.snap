// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopEntitiesViewer should render correctly 1`] = `
<WithStyles(ForwardRef(Table))>
  <WithStyles(ForwardRef(TableBody))>
    <InspectTooltip
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      key="0"
      open={true}
      title={
        <div>
          <div>
            Resource:
            Test
          </div>
          <ForwardRef(WithStyles) />
          <div>
            Scope and Usage:
          </div>
          <p>
            Verbose details
          </p>
        </div>
      }
    >
      <WithStyles(ForwardRef(TableRow))
        hover={true}
        key="Test"
        onClick={[Function]}
        selected={true}
      >
        <WithStyles(ForwardRef(TableCell))
          component="th"
          scope="row"
        >
          <span
            className="makeStyles-span-1"
            onClick={[Function]}
          >
            Test
          </span>
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </InspectTooltip>
  </WithStyles(ForwardRef(TableBody))>
</WithStyles(ForwardRef(Table))>
`;
